packages/vats/src/core/promise-space.js

```diff
@@ -95,86 +95,81 @@ export const makePromiseSpace = (optsOrLog = {}) => {
-    } else {
+    let state = nameToState.get(name);
+    if (!state) {
       onAddKey(name);
+      remaining.add(name);
       const pk = makePromiseKit();
-
       pk.promise
         .finally(() => {
-          remaining.delete(name);
           onSettled(name, remaining);
         })
         .catch(() => {});
-
-      const settling = () => {
-        assert(state);
-        state = harden({ ...state, isSettling: true });
-        nameToState.set(name, state);
-      };
-
-      const resolve = value => {
-        settling();
-        onResolve(name, value);
-        pk.resolve(value);
-      };
-      const reject = reason => {
-        settling();
-        pk.reject(reason);
-      };
-
-      const reset = (reason = undefined) => {
-        onReset(name);
-        if (!state.isSettling) {
-          if (!reason) {
-            // Reuse the old promise; don't reject it.
-            return;
-          }
-          reject(reason);
-        }
-        // Now publish a new promise.
-        nameToState.delete(name);
-        remaining.delete(name);
-      };
-
-      state = harden({
-        isSettling: false,
-        resolve,
-        reject,
-        reset,
-        promise: pk.promise,
-      });
+      state = harden({ pk, isSettling: false });
       nameToState.set(name, state);
-      remaining.add(name);
     }
     return state;
   };
 
+  // we must tolerate these producer methods being retrieved both
+  // before and after the consumer is retrieved, and also both before
+  // and after reset() is invoked, so they only close over 'name' and
+  // not over any state variables
+
+  const makeProducer = name => {
+    const resolve = value => {
+      onResolve(name, value);
+      const old = provideState(name);
+      nameToState.set(name, harden({ ...old, isSettling: true }));
+      old.pk.resolve(value);
+      remaining.delete(name);
+    };
+    const reject = reason => {
+      const old = provideState(name);
+      nameToState.set(name, harden({ ...old, isSettling: true }));
+      old.pk.reject(reason);
+      remaining.delete(name);
+    };
+    const reset = (reason = undefined) => {
+      onReset(name);
+      const old = provideState(name);
+      if (!old.isSettling) {
+        // we haven't produced a value yet, and there might be
+        // consumers still watching old.pk.promise
+        if (reason === undefined) {
+          // so just let them wait for the new value: resetting an
+          // unresolved item is a no-op
+          return;
+        }
+        // reject those watchers; new watchers will wait for the new
+        // value through the replacement promise
+        reject(reason);
+      }
+      // delete the state, so new callers will get a new promise kit
+      nameToState.delete(name);
+      remaining.delete(name);
+    };
+
+    return harden({ resolve, reject, reset });
+  };
+
   /** @type {PromiseSpaceOf<T>['consume']} */
   // @ts-expect-error cast
   const consume = new Proxy(
     {},
```

Comment:
```suggestion
    harden({}),
```
This prevents non-get operations from mutating the target and using it as a communications channel. Should be done for both proxies.

Reply:
> @erights You mentioned that the Proxy could be safer, but it wasn't obvious to me what should be done.. let me know what should be changed.

That's what I had in mind.

Reply:
Btw, I know it is strange that the proxy invariants allow `get` to pretend properties exist when the hardened target doesn't have them. That's because the proxy invariants are only strong on own-invariants and weak on inherited-only invariants. The proxy can hallucinate these properties only because they are absent as own properties on the target.

Ask me about the history here sometime if your curious. We landed in an ok but weirder than necessary place.

Reply:
You need to `yarn format` and push again